compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -1112,8 +1125,48 @@ pub fn typeid_for_instance<'tcx>(
     mut instance: Instance<'tcx>,
     options: TypeIdOptions,
 ) -> String {
-    if matches!(instance.def, ty::InstanceDef::Virtual(..)) {
-        instance.args = strip_receiver_auto(tcx, instance.args)
+    if (matches!(instance.def, ty::InstanceDef::Virtual(..))
+        && instance.def_id() == tcx.lang_items().drop_in_place_fn().unwrap())
+        || matches!(instance.def, ty::InstanceDef::DropGlue(_, Some(_ty)))
+    {
+        // Adjust the type ids of DropGlues
+        //
+        // DropGlues may have indirect calls to one or more given types drop function. Rust allows
+        // for types to be erased to any trait object and retains the drop function for the original
+        // type, which means at the indirect call sites in DropGlues, when typeid_for_fnabi is
+        // called a second time, it only has information after type erasure and it could be a call
+        // on any arbitrary trait object. Normalize them to a synthesized Drop trait object, both on
+        // declaration/definition, and during code generation at call sites so they have the same
+        // type id and match.
+        //
+        // FIXME(rcvalle): This allows a drop call on any trait object of types that have a
+        //   non-empty DropGlue to call the DropGlue of any other types that have a non-empty
+        //   DropGlue.
+        let fn_abi = tcx
+            .fn_abi_of_instance(tcx.param_env(instance.def_id()).and((instance, ty::List::empty())))
+            .unwrap_or_else(|instance| {
+                bug!("typeid_for_instance: couldn't get fn_abi of instance {:?}", instance)
+            });
+        let self_ty = if fn_abi.args[0].layout.ty.is_mutable_ptr() {
+            Ty::new_mut_ref(
+                tcx,
+                tcx.lifetimes.re_erased,
+                new_dynamic_trait(tcx, tcx.lang_items().drop_trait().unwrap(), List::empty()),
```

Comment:
Since this contracts everything to a single type anyways, you could do `*mut ()` rather than `&mut dyn Drop`. It'd be simpler to write.

Reply:
I think it's better to use the explicit trait object to avoid grouping it with something else. It also helps with debugging and when looking at type metadata in .ll files.